workers/unity/Packages/io.improbable.gdk.mobile/Editor/MobileLauncherWindow.cs

```diff
@@ -76,84 +130,107 @@ public void OnGUI()
+
+            var names = isAndroid
+                ? isEmulator
+                    ? androidEmulatorNames
+                    : androidDeviceNames
+                : isEmulator
+                    ? iOSSimulatorNames
+                    : iOSDeviceNames;
+
+            var devices = isAndroid
+                ? isEmulator
+                    ? androidEmulators
+                    : androidDevices
+                : isEmulator
+                    ? iOSSimulators
+                    : iOSDevices;
+
+            // List of devices/emulators with refresh button linked to correct platform
+            using (new GUILayout.HorizontalScope())
+            {
+                using (new EditorGUI.DisabledScope(names.Length == 0))
+                {
+                    index = EditorGUILayout.Popup($"{deviceOrEmulator} Model", index, names);
                 }
 
-                using (new EditorGUI.DisabledScope(simulatorNames.Length == 0))
+                var buttonIcon = new GUIContent(EditorGUIUtility.IconContent("Refresh"))
                 {
-                    if (GUILayout.Button("Launch iOS app in Simulator"))
+                    tooltip = $"Refresh your {(!isAndroid && isEmulator ? "Simulator" : deviceOrEmulator.ToLower())} list."
+                };
+
+                if (GUILayout.Button(buttonIcon, EditorStyles.miniButton, GUILayout.ExpandWidth(false)))
+                {
+                    if (isAndroid)
                     {
-                        var availableSimulators = iOSLaunchUtils.RetrieveAvailableiOSSimulators();
-                        if (availableSimulators.TryGetValue(simulatorNames[simulatorNameIndex], out var simulatorUID))
-                        {
-                            iOSLaunchUtils.Launch(launchConfig.ShouldConnectLocally, simulatorUID, launchConfig.RuntimeIp, true);
-                        }
-                        else
-                        {
-                            simulatorNames = availableSimulators.Keys.ToArray();
-                            simulatorNameIndex = 0;
-                            Debug.LogError("Failed to launch app on selected simulator.");
-                        }
+                        RefreshAndroidEmulatorsAndDevices();
+                    }
+                    else
+                    {
+                        RefreshiOSEmulatorsAndDevices();
                     }
                 }
+            }
 
-                CommonUIElements.DrawHorizontalLine(8, DarkGrey);
+            // Draw a launch button that will launch the app on the correct platform and chosen device
+            using (new EditorGUI.DisabledScope(names.Length == 0))
+            {
+                if (!GUILayout.Button($"Launch {androidOriOs} app on {deviceOrEmulator}"))
+                {
+                    return;
+                }
 
-                using (new GUILayout.HorizontalScope())
+                if (devices.TryGetValue(names[index], out var deviceId))
```

Comment:
If we stored an object here instead of an ID, we might be able to do something like: 

```csharp
if (devices.TryGetValue(names[index], out var device)
{
	device.Launch(options);
}
```

and delegate implementation to the object.